kubernetes/client/ws_client.py

```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
         else:
             ssl_opts = {'cert_reqs': ssl.CERT_NONE}
 
-        self.ws.run_forever(sslopt=ssl_opts)
-
-    def on_message(self, ws, message):
-        if message[0] == '\x01':
-            message = message[1:]
-        if message:
-            if six.PY3 and isinstance(message, six.binary_type):
-                message = message.decode('utf-8')
-            self.messages.append(message)
-
-    def on_error(self, ws, error):
-        self.errors.append(error)
-
-    def on_close(self, ws):
-        pass
-
-    def on_open(self, ws):
-        pass
+        self.sock = WebSocket(sslopt=ssl_opts, skip_utf8_validation=False)
+        self.sock.connect(url, header=header)
+        self._connected = True
+
+    def peek_stdout(self):
+        self.update()
+        return self._stdout
+
+    def read_stdout(self):
+        if not self._stdout:
+            self.update(timeout=None)
+        ret = self._stdout
+        self._stdout = ""
+        return ret
+
+    def peek_stderr(self):
+        self.update()
+        return self._stderr
+
+    def read_stderr(self):
+        if not self._stderr:
+            self.update(timeout=None)
+        ret = self._stderr
+        self._stderr = ""
+        return ret
+
+    def read_all(self):
+        out = self._all
+        self._all = ""
+        self._stdout = ""
+        self._stderr = ""
+        return out
+
+    def is_open(self):
+        return self._connected
+
+    # TODO: This method does not seem to work.
+    def write_stdin(self, data):
+        self.sock.send("\x00" + data)
+
+    def update(self, timeout=0):
+        if not self.is_open():
+            return
+        if not self.sock.connected:
+            self._connected = False
+            return
+        r, _, _ = select.select(
+            (self.sock.sock, ), (), (), timeout)
+        if r:
+            op_code, frame = self.sock.recv_data_frame(True)
+            if op_code == ABNF.OPCODE_CLOSE:
+                self._connected = False
+                return
+            elif op_code == ABNF.OPCODE_BINARY or op_code == ABNF.OPCODE_TEXT:
+                data = frame.data
+                if six.PY3:
+                    data = data.decode("utf-8")
+                if data[0] == '\x01':
+                    data = data[1:]
+                if data:
+                    self._all += data
+                    if data[0] == '\x02':
+                        self._stderr += data[1:]
+                    else:
+                        self._stdout += data
+
+    def run_forever(self, timeout=None):
+        if timeout:
```

Comment:
Can we get an explanation on what run_forever means and are there any gotchas(what are the assumptions you are making)?